conformance/tests/gateway-dynamic-listeners.go

```diff
@@ -0,0 +1,223 @@
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+	"sigs.k8s.io/gateway-api/conformance/utils/tls"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, GatewayListenerDynamicPorts)
+}
+
+var GatewayListenerDynamicPorts = suite.ConformanceTest{
+	ShortName: "GatewayListenerDynamicPorts",
+	Features: []suite.SupportedFeature{
+		suite.SupportGateway,
+		suite.SupportGatewayListenerDynamicPorts,
+	},
+	Description: "A Gateway in the gateway-conformance-infra namespace should handle adding and removing listeners with arbitrary ports",
+	Manifests:   []string{"tests/gateway-dynamic-listeners.yaml"},
+	Test: func(t *testing.T, s *suite.ConformanceTestSuite) {
+
+		// Ephemeral port range
+		const (
+			portCount    = 10
+			tlsPortCount = 5
+		)
+		var (
+			portStart = s.ListenerConfig.DynamicPortRange.Start
+			portEnd   = s.ListenerConfig.DynamicPortRange.End
+
+			gwNN       = types.NamespacedName{Name: "gateway-dynamic-listener", Namespace: "gateway-conformance-infra"}
+			namespaces = []string{"gateway-conformance-infra"}
+			certNN     = types.NamespacedName{Name: "tls-wildcard-hostname", Namespace: gwNN.Namespace}
+			ports      = sets.New[int]()
+			listeners  = make([]v1beta1.Listener, 0, portCount)
+			same       = v1beta1.NamespacesFromSame
+
+			expectedListeners = []v1beta1.ListenerStatus{{
+				Name: "http",
+				SupportedKinds: []v1beta1.RouteGroupKind{{
+					Group: (*v1beta1.Group)(&v1beta1.GroupVersion.Group),
+					Kind:  v1beta1.Kind("HTTPRoute"),
+				}},
+				Conditions: []metav1.Condition{{
+					Type:   string(v1beta1.ListenerConditionAccepted),
+					Status: metav1.ConditionTrue,
+					Reason: "", //any reason
+				}},
+				AttachedRoutes: 1,
+			}}
+		)
+		if portEnd < portStart {
+			t.Fatal("DynamicPortRange.Start must be less than DynamicPortRange.End")
+		}
+
+		if portEnd-portStart < portCount {
+			t.Fatal("DynamicPortRange input requires at least 10 ports")
+		}
+
+		for i := 0; i < portCount; i++ {
+			port := nextPort(portStart, portEnd, ports)
+
+			listeners = append(listeners, v1beta1.Listener{
+				Name:     v1beta1.SectionName(strconv.Itoa(port)),
+				Port:     v1beta1.PortNumber(port),
+				Protocol: v1beta1.HTTPProtocolType,
+				AllowedRoutes: &v1beta1.AllowedRoutes{
+					Namespaces: &v1beta1.RouteNamespaces{From: &same},
+				},
+			})
+
+			expectedListeners = append(expectedListeners, v1beta1.ListenerStatus{
+				Name: v1beta1.SectionName(strconv.Itoa(port)),
+				SupportedKinds: []v1beta1.RouteGroupKind{{
+					Group: (*v1beta1.Group)(&v1beta1.GroupVersion.Group),
+					Kind:  v1beta1.Kind("HTTPRoute"),
```

Comment:
I don't think we want to assume that this is the only Route type that an implementation may support with the "HTTP" protocol.

Reply:
closing as dupe - let's talk about it in this thread https://github.com/kubernetes-sigs/gateway-api/pull/1859#discussion_r1234153687